### PR TITLE
Fix for users receiving a mustAuth response from the backend even though they set the password in this session.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
If a user has set the password in this session (either dialog or main site) and they receive a mustAuth response from the verification poll, authenticate the user with the already entered email address and password.

This fixes users having to enter their password a second time on the /forgot and /signup pages after the verification is complete.

issue #1682

### CODE REVIEW

File: resources/static/common/js/user.js [hunk 1]
```diff
@@ -103,17 +105,26 @@ BrowserID.User = (function() {
...
-    // RP and logged in.
-    var site = User.getReturnTo();
-    if (staged && site) storage.setReturnTo(site);
-
-    complete(onComplete, status);
+  function stageAddressVerification(email, password, stagingStrategy, onComplete, onFailure) {
+    // These are saved for the addressVerificationPoll.  If there is
+    // a stagedEmail or stagedPassword when the poll completes, try to
+    // authenticate the user.
+    stagedEmail = email;
```
Comment (shane-tomlinson) [inline resources/static/common/js/user.js:110]: Every time an email/password is staged, save off the staged email and password to use on completion.

File: resources/static/common/js/user.js [hunk 2]
```diff
@@ -148,6 +159,68 @@ BrowserID.User = (function() {
   }
 
   function addressVerificationPoll(checkFunc, email, onSuccess, onFailure) {
+    function userVerified(completionStatus) {
+      if (stagedEmail && stagedPassword) {
```
Comment (shane-tomlinson) [inline resources/static/common/js/user.js:161]: If there is a stagedEmail and stagedPassword, re-authenticate the user using the stagedEmail and stagedPassword.

File: resources/static/common/js/user.js [hunk 3]
```diff
@@ -148,6 +159,68 @@ BrowserID.User = (function() {
...
+          completeVerification(completionStatus);
+        }, onFailure);
+      }
+    }
+
+    function completeVerification(status) {
+      // As soon as the registration comes back as complete, we should
+      // ensure that the stagedOnBehalfOf is cleared so there is no stale
+      // data.
+      storage.setReturnTo("");
```
Comment (shane-tomlinson) [inline resources/static/common/js/user.js:195]: This logic is taken verbatim from below.

**Comment by shane-tomlinson** [comment]
Note that once this is merged, the old code in resources/static/dialog/js/modules/check_registration can be removed.

@lloyd - can you have a look at this?

**Comment by seanmonstar** [comment]
@shane-tomlinson since the line comments seem needed to explain what's going on, should they be in the code instead?

**Comment by shane-tomlinson** [comment]
@seanmonstar - outside of the last comment about "this logic is taken verbatim from below", the comments are already in the code as well, and slightly more verbose. I wanted to draw attention to them in the pull request.

**Comment by seanmonstar** [comment]
This looks fine to me, the only thing stopping me from clicking merge is that it's 3 months old. I can click?

**Pull request merged**